pgjdbc/src/main/java/org/postgresql/util/PasswordUtil.java

```diff
@@ -0,0 +1,189 @@
+import com.ongres.scram.common.bouncycastle.base64.Base64;
+import com.ongres.scram.common.stringprep.StringPreparations;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.nio.charset.StandardCharsets;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.sql.SQLException;
+
+public class PasswordUtil {
+  private static final String DEFAULT_PASSWORD_ENCRYPTION = "scram-sha-256";
+  private static final int DEFAULT_ITERATIONS = 4096;
+  private static final int DEFAULT_SALT_LENGTH = 16;
+
+  private static class SecureRandomHolder {
+    static final SecureRandom INSTANCE = new SecureRandom();
+  }
+
+  private static SecureRandom getSecureRandom() {
+    return SecureRandomHolder.INSTANCE;
+  }
+
+  /**
+   * Generate the encoded text representation of the given password for
+   * SCRAM-SHA-256 authentication. The return value of this method is the literal
+   * text that may be used when creating or modifying a user with the given
+   * password without the surrounding single quotes.
+   *
+   * @param password   The plain text of the user's password
+   * @param iterations The number of iterations of the hashing algorithm to
+   *                   perform
+   * @param salt       The random salt value
+   * @return T text representation of the password encrypted for SCRAM-=SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password, int iterations, byte[] salt) {
+    if (iterations <= 0) {
+      throw new IllegalArgumentException("iterations must be greater than zero");
+    }
+    if (salt.length == 0) {
+      throw new IllegalArgumentException("salt length must be greater than zero");
+    }
+    String passwordText = String.valueOf(password);
+    byte[] saltedPassword = ScramFunctions.saltedPassword(ScramMechanisms.SCRAM_SHA_256,
+        StringPreparations.SASL_PREPARATION, passwordText, salt, iterations);
+    byte[] clientKey = ScramFunctions.clientKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+    byte[] storedKey = ScramFunctions.storedKey(ScramMechanisms.SCRAM_SHA_256, clientKey);
+    byte[] serverKey = ScramFunctions.serverKey(ScramMechanisms.SCRAM_SHA_256, saltedPassword);
+
+    return "SCRAM-SHA-256" //
+      + "$" + iterations //
+      + ":" + Base64.toBase64String(salt) //
+      + "$" + Base64.toBase64String(storedKey) //
+      + ":" + Base64.toBase64String(serverKey);
+  }
+
+  /**
+   * Encode the given password for SCRAM-SHA-256 authentication using the default
+   * iteration count and a random salt.
+   *
+   * @param password The plain text of the user's password
+   * @return The text representation of the password encrypted for SCRAM-=SHA-256
+   *         authentication
+   */
+  public static String encodeScramSha256(char[] password) {
+    SecureRandom rng = getSecureRandom();
+    byte[] salt = rng.generateSeed(DEFAULT_SALT_LENGTH);
```

Comment:
SecureRandom is not thread-safe, so synchronization is needed here

Reply:
SecureRandom _is_ thread safe and no synchronization is required. It inherits from Random which is declared as thread-safe in the docs: https://docs.oracle.com/javase/8/docs/api/java/util/Random.html

There might be some contention if many concurrent threads were invoking that method but that's hardly a realistic use case. It's not for authentication, it's for generating a salt for encoding a password so it's unlikely to ever be run concurrently. And even if it does it'll just be a bit slow.

Reply:
My bad. Thank you for checking that